src/lsm/block_free_set.zig

```diff
@@ -171,6 +200,66 @@ test "BlockFreeSet acquire/release" {
     try test_acquire_release(65 * BlockFreeSet.shard_size);
 }
 
+test "BlockFreeSet checkpoint" {
+    const expectEqual = std.testing.expectEqual;
+    const blocks_count = BlockFreeSet.shard_size;
+    var set = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer set.deinit(std.testing.allocator);
+
+    var empty = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer empty.deinit(std.testing.allocator);
+
+    var full = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer full.deinit(std.testing.allocator);
+
+    var i: usize = 0;
+    while (i < full.blocks.bit_length) : (i += 1) {
+        try expectEqual(@as(?u64, i + 1), full.acquire());
+    }
+
+    {
+        // Allocate & stage-release every block.
+        i = 0;
+        while (i < set.blocks.bit_length) : (i += 1) {
+            try expectEqual(@as(?u64, i + 1), set.acquire());
+            set.release_at_next_checkpoint(i + 1);
+        }
+        // All blocks are still allocated, though staged to release at the next checkpoint.
+        try expectEqual(@as(?u64, null), set.acquire());
+    }
+
+    // Free all the blocks.
+    set.checkpoint();
+    try expect_block_free_set_equal(empty, set);
+
+    // Redundant checkpointing is a noop (but safe).
+    set.checkpoint();
+
+    // Allocate & stage-release all blocks again.
```

Comment:
Nice test! These short "what" explanation comments are great for tests. To give a quick overview and tell a story of what the test is doing. It makes it easier for readers to tell at a glance what the test is aiming to do. I've been thinking we should do this more and more for tests, also with a short summary at the top for other tests of the big idea or test strategy.